openquake/hazardlib/near_fault.py

```diff
@@ -0,0 +1,476 @@
+"""
+Module :mod:`openquake.hazardlib.nearfault` provides methods for near fault
+PSHA calculation.
+"""
+
+import math
+import numpy as np
+from openquake.hazardlib.geo import geodetic as geod
+import scipy.spatial.distance as dst
+
+
+def get_xyz_from_ll(projected, reference):
+    """
+    This method computes the x ,y and z coordinates of a set of points
+    provided a reference point
+
+    :param projected:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the coordinates of target point to be projected
+    :param reference:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the coordinates of the reference point.
+
+    :returns:
+            x
+            y
+            z
+    """
+
+    azims = geod.azimuth(reference.longitude, reference.latitude,
+                         projected.longitude, projected.latitude)
+    depths = np.subtract(reference.depth, projected.depth)
+    dists = geod.geodetic_distance(reference.longitude,
+                                   reference.latitude,
+                                   projected.longitude,
+                                   projected.latitude)
+    return (dists * math.sin(math.radians(azims)),
+            dists * math.cos(math.radians(azims)),
+            depths)
+
+
+def get_plane_equation(p0, p1, p2, reference):
+    '''
+    Define plane equation by 3 given points, p0, p1, p2, which format in
+    class:`~openquake.hazardlib.geo.point.Point` object
+
+    :param p0:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of the one vertex of the fault patch.
+    :param p1:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of the one vertex of the fault patch.
+        Note: the order of the vertex should give
+        in clockwise.
+    :param p2:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of the one vertex of the fault patch.
+        Note: the order of the vertex should give
+        in clockwise.
+    :param reference:
+        :class:`~openquake.hazardlib.geo.point.Point` object
+        representing the location of project reference point
+    :returns:
+        normal: normal vector of the plane (a,b,c)
+        dist_to_plane: d in the plane equation, ax + by + cz = d
+
+    '''
+
+    p0_xyz = get_xyz_from_ll(p0, reference)
+    p1_xyz = get_xyz_from_ll(p1, reference)
+    p2_xyz = get_xyz_from_ll(p2, reference)
+
+    p0 = np.array(p0_xyz)
+    p1 = np.array(p1_xyz)
+    p2 = np.array(p2_xyz)
+
+    u = p1 - p0
+    v = p2 - p0
+
+    # vector normal to plane, ax+by+cy = d, normal=(a,b,c)
+
+    normal = np.cross(u, v)
+
+    # Define the d for the plane equation
+    dist_to_plane = np.dot(p0, normal)
+
+    return normal, dist_to_plane
+
+
+def projection_pp(site, normal, dist_to_plane, reference):
+    '''
+    Find the projection point Pp on the patch plane
```

Comment:
This method finds the point on a fault patch defined as the Pp (i.e. 'perpendicular projection of PS onto the fault plane' Spudich et al. (2013) - page 88) given a site.